### PR TITLE
Add FXIOS-7892 [v123] Display a progress bar while (re)analysis is in progress

### PR DESCRIPTION
**Pull request by PARAIPAN9**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-7892)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/17635)

## :bulb: Description

- Implement circular progress view for tracking the analysis progress of a product.
### Demo

https://github.com/mozilla-mobile/firefox-ios/assets/51127880/69fafc38-914e-407d-be85-375ec149a004

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [x] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed I updated documentation / comments for complex code and public methods

### CODE REVIEW

File: firefox-ios/Client/Frontend/Fakespot/Views/FakespotMessageCardView.swift [hunk 1]
```diff
@@ -194,25 +196,45 @@ final class FakespotMessageCardView: UIView, ThemeApplicable, Notifiable {
         fatalError("init(coder:) has not been implemented")
     }
 
-    func configure(_ viewModel: FakespotMessageCardViewModel) {
+    func configure(
+        _ viewModel: FakespotMessageCardViewModel,
+        fakespotViewModel: FakespotViewModel? = nil
```
Comment (razvanlitianu) [inline firefox-ios/Client/Frontend/Fakespot/Views/FakespotMessageCardView.swift:201]: I'm not really comfortable with the idea of passing 2 viewModels. I think we should have a way to update the view if the progress change.

File: firefox-ios/Client/Frontend/Fakespot/FakespotViewModel.swift [hunk 2]
```diff
@@ -167,6 +168,9 @@ class FakespotViewModel {
         }
     }
 
+    @Published var analysisProgress: Double = 0
+    var analysisProgressChanged: ((Double) -> Void)?
```
[the remark this answers is not in the document]
Comment (PARAIPAN9) [reply firefox-ios/Client/Frontend/Fakespot/FakespotViewModel.swift:172]: Totally agree, but taking into consideration the susceptibility of using Combine, I opted for a closure instead.

File: firefox-ios/Client/Frontend/Strings.swift [hunk 3]
```diff
@@ -5938,6 +5938,15 @@ extension String {
                 comment: "Label text that gets presented as a confirmation at the bottom of screen when credit card information gets updated successfully",
                 lastUsedInVersion: 121)
         }
+
+        struct v123 {
```
Comment (thatswinnie) [inline firefox-ios/Client/Frontend/Strings.swift:5942]: The string was last used in 122. This PR is for 123.

File: firefox-ios/Client/Frontend/Fakespot/FakespotViewModel.swift [hunk 4]
```diff
@@ -5,6 +5,7 @@
 import UIKit
 import Common
 import Shared
+import SwiftUI
```
Comment (thatswinnie) [inline firefox-ios/Client/Frontend/Fakespot/FakespotViewModel.swift:8]: Is this import still needed here?

File: firefox-ios/Client/Frontend/Fakespot/Views/FakespotMessageCardView.swift [hunk 5]
```diff
@@ -198,21 +238,35 @@ final class FakespotMessageCardView: UIView, ThemeApplicable, Notifiable {
...
             let imageView: UIImageView = .build { imageView in
                 imageView.contentMode = .scaleAspectFit
                 imageView.image = UIImage(named: name)
             }
             accessoryView = imageView
         case .progress:
-            let spinner = UIActivityIndicatorView(style: .medium)
-            spinner.startAnimating()
-            accessoryView = spinner
+            titleLabel.text = String(format: viewModel.title, viewModel.formatProgress())
```
Comment (thatswinnie) [inline firefox-ios/Client/Frontend/Fakespot/Views/FakespotMessageCardView.swift:253]:
```suggestion
            titleLabel.text = String.localizedStringWithFormat(viewModel.title, viewModel.formatProgress())
```

File: firefox-ios/Client/Frontend/Fakespot/Views/FakespotMessageCardView.swift [hunk 6]
```diff
@@ -198,21 +238,35 @@ final class FakespotMessageCardView: UIView, ThemeApplicable, Notifiable {
...
+                size: UX.progressViewFontSize
+            )
+
+            viewModel.analysisProgressChanged = { [weak self] progress in
+                let progressLevel = viewModel.formatProgress(progress / 100.0)
+                self?.titleLabel.text = String(format: viewModel.title, progressLevel)
+            }
+
+            let swiftUICircularProgressView = CircularProgressView(viewModel: viewModel)
+            let circularProgressView = UIHostingController(rootView: swiftUICircularProgressView).view!
```
Comment (thatswinnie) [inline firefox-ios/Client/Frontend/Fakespot/Views/FakespotMessageCardView.swift:265]: Let's not force-unwrap the view.

**Review by thatswinnie — APPROVED** [review]

**Pull request merged**